unified_planning/io/ma_pddl_writer.py

```diff
@@ -0,0 +1,1066 @@
+    _update_domain_objects,
+    ConverterToPDDLString,
+    PDDL_KEYWORDS,
+    INITIAL_LETTER,
+    _get_pddl_name,
+)
+
+from unified_planning.model.walkers.simplifier import Simplifier
+from unified_planning.model.expression import ExpressionManager, Expression
+from unified_planning.model.walkers.substituter import Substituter
+import re
+
+
+class ConverterToMAPDDLString(ConverterToPDDLString):
+    """Expression converter to a MA-PDDL string."""
+
+    def __init__(
+        self,
+        env: "up.environment.Environment",
+        get_mangled_name: Callable[
+            [
+                Union[
+                    "up.model.Type",
+                    "up.model.Action",
+                    "up.model.Fluent",
+                    "up.model.Object",
+                    "up.model.multi_agent.Agent",
+                ]
+            ],
+            str,
+        ],
+    ):
+        ConverterToPDDLString.__init__(self, env, get_mangled_name)
+
+    def walk_dot(self, expression, args):
+        agent = expression.agent()
+        fluent = expression.args[0].fluent()
+        objects = expression.args[0].args
+        # f.args[0].fluent(), f.agent().name, f.args[0].args[0]]
+        return f'({self.get_mangled_name(fluent)}{" "}{self.get_mangled_name(agent)}{" " if len(args) > 0 else ""}{" ".join([self.convert(obj) for obj in objects])})'
+
+
+class MAPDDLWriter:
+    """This class can be used to write a :class:`~unified_planning.model.MultiAgentProblem` in `MA-PDDL`."""
+
+    def __init__(
+        self,
+        problem: "up.model.multi_agent.MultiAgentProblem",
+        needs_requirements: bool = True,
+    ):
+        self._env = problem.env
+        self.problem = problem
+        self.problem_kind = self.problem.kind
+        self.needs_requirements = needs_requirements
+        # otn represents the old to new renamings
+        self.otn_renamings: Dict[
+            Union[
+                "up.model.Type",
+                "up.model.Action",
+                "up.model.Fluent",
+                "up.model.Object",
+                "up.model.Parameter",
+                "up.model.Variable",
+                "up.model.multi_agent.Agent",
+            ],
+            str,
+        ] = {}
+        # nto represents the new to old renamings
+        self.nto_renamings: Dict[
+            str,
+            Union[
+                "up.model.Type",
+                "up.model.Action",
+                "up.model.Fluent",
+                "up.model.Object",
+                "up.model.Parameter",
+                "up.model.Variable",
+                "up.model.multi_agent.Agent",
+            ],
+        ] = {}
+        # those 2 maps are "simmetrical", meaning that "(otn[k] == v) implies (nto[v] == k)"
+        self.domain_objects: Optional[Dict[_UserType, Set[Object]]] = None
+        self.domain_objects_agents: Dict[up.model.multi_agent.Agent, str]
+        self.domain_fluents_agents: Dict[up.model.multi_agent.Agent, List[Fluent]]
```

Comment:
What is the purpose of this state? The Printer class should ideally be reusable across multiple problems and adding state requires cleanup (that seems to be missing)

Reply:
The states `domain_objects_agents` and `domain_fluents_agents` are used to keep track of fluents and objects of other agents that are referenced using the DOT operator in the preconditions. They allow the addition of these fluents and objects to the public predicates and constants in the MA-PDDL domain of the current agent. This is necessary since these fluents are not present in the current agent, and the current agent's fluents do not use the objects.